File: lib/analysis/stft.dart
import 'dart:math' as math;
import 'dart:typed_data';

import '../playback/audio_levels.dart';

/// STFT → log-folded 20-band spectrum frames.
class StftBandFolder {
  StftBandFolder({
    this.fftSize = 1024,
    this.framesPerSecond = 30,
  });

  final int fftSize;
  final int framesPerSecond;

  /// Analyse mono PCM into per-frame band energies (0..1).
  List<List<double>> analyze(
    Float64List samples, {
    required int sampleRateHz,
  }) {
    if (samples.isEmpty || sampleRateHz <= 0) return const [];

    final hop = math.max(1, sampleRateHz ~/ framesPerSecond);
    final window = _hann(fftSize);
    final bandEdges = _logBandEdges(sampleRateHz, fftSize);
    final frames = <List<double>>[];

    for (var start = 0; start + fftSize <= samples.length; start += hop) {
      final re = Float64List(fftSize);
      final im = Float64List(fftSize);
      for (var i = 0; i < fftSize; i++) {
        re[i] = samples[start + i] * window[i];
      }
      _fftInPlace(re, im);

      final mags = Float64List(fftSize ~/ 2);
      for (var k = 0; k < mags.length; k++) {
        mags[k] = math.sqrt(re[k] * re[k] + im[k] * im[k]);
      }

      final bands = List<double>.filled(AudioLevels.bandCount, 0);
      for (var b = 0; b < AudioLevels.bandCount; b++) {
        final lo = bandEdges[b];
        final hi = bandEdges[b + 1];
        var sum = 0.0;
        var count = 0;
        for (var k = lo; k < hi && k < mags.length; k++) {
          sum += mags[k];
          count++;
        }
        bands[b] = count == 0 ? 0 : sum / count;
      }

      // Normalise per frame so a loud tone fills the display.
      var peak = 0.0;
      for (final v in bands) {
        if (v > peak) peak = v;
      }
      if (peak > 1e-12) {
        for (var i = 0; i < bands.length; i++) {
          bands[i] = (bands[i] / peak).clamp(0.0, 1.0);
        }
      }
      frames.add(bands);
    }

    if (frames.isEmpty) {
      frames.add(List<double>.filled(AudioLevels.bandCount, 0));
    }
    return frames;
  }

  /// Centre frequency (Hz) of display band [index].
  static double bandCenterHz(int index, {required int sampleRateHz}) {
    final edgesHz = _logBandEdgeHz(sampleRateHz);
    return math.sqrt(edgesHz[index] * edgesHz[index + 1]);
  }

  static List<int> _logBandEdges(int sampleRateHz, int fftSize) {
    final edgesHz = _logBandEdgeHz(sampleRateHz);
    final binHz = sampleRateHz / fftSize;
    return [
      for (final hz in edgesHz)
        (hz / binHz).floor().clamp(1, (fftSize ~/ 2) - 1),
    ];
  }

  static List<double> _logBandEdgeHz(int sampleRateHz) {
    const fMin = 40.0;
    final fMax = sampleRateHz / 2.0;
    final edges = <double>[];
    for (var i = 0; i <= AudioLevels.bandCount; i++) {
      final t = i / AudioLevels.bandCount;
      edges.add(fMin * math.pow(fMax / fMin, t));
    }
    return edges;
  }

  static Float64List _hann(int n) {
    final w = Float64List(n);
    for (var i = 0; i < n; i++) {
      w[i] = 0.5 * (1 - math.cos(2 * math.pi * i / (n - 1)));
    }
    return w;
  }

  /// In-place radix-2 Cooley–Tukey FFT.
  static void _fftInPlace(Float64List re, Float64List im) {
    final n = re.length;
    if (n == 0 || (n & (n - 1)) != 0) {
      throw ArgumentError('fftSize must be power of two');
    }

    // Bit-reverse permutation.
    var j = 0;
    for (var i = 1; i < n; i++) {
      var bit = n >> 1;
      for (; j & bit != 0; bit >>= 1) {
        j ^= bit;
      }
      j ^= bit;
      if (i < j) {
        final tr = re[i];
        re[i] = re[j];
        re[j] = tr;
        final ti = im[i];
        im[i] = im[j];
        im[j] = ti;
      }
    }

    for (var len = 2; len <= n; len <<= 1) {
      final ang = -2 * math.pi / len;
      final wlenRe = math.cos(ang);
      final wlenIm = math.sin(ang);
      for (var i = 0; i < n; i += len) {
        var wRe = 1.0;
        var wIm = 0.0;
        for (var k = 0; k < len ~/ 2; k++) {
          final uRe = re[i + k];
          final uIm = im[i + k];
          final vRe = re[i + k + len ~/ 2] * wRe - im[i + k + len ~/ 2] * wIm;
          final vIm = re[i + k + len ~/ 2] * wIm + im[i + k + len ~/ 2] * wRe;
          re[i + k] = uRe + vRe;
          im[i + k] = uIm + vIm;
          re[i + k + len ~/ 2] = uRe - vRe;
          im[i + k + len ~/ 2] = uIm - vIm;
          final nextWRe = wRe * wlenRe - wIm * wlenIm;
          wIm = wRe * wlenIm + wIm * wlenRe;
          wRe = nextWRe;
        }
      }
    }
  }
}

File: src/stft.cpp
#include "stft.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <vector>

namespace aoide {
namespace {

constexpr double kPi = 3.14159265358979323846;

std::vector<double> hann(int n) {
  std::vector<double> w(size_t(n), 0.0);
  for (int i = 0; i < n; ++i) {
    w[size_t(i)] = 0.5 * (1.0 - std::cos(2.0 * kPi * double(i) / double(n - 1)));
  }
  return w;
}

std::vector<double> logBandEdgeHz(int sampleRateHz) {
  constexpr double fMin = 40.0;
  const double fMax = sampleRateHz / 2.0;
  std::vector<double> edges;
  edges.reserve(AudioLevels::kBandCount + 1);
  for (int i = 0; i <= AudioLevels::kBandCount; ++i) {
    const double t = double(i) / double(AudioLevels::kBandCount);
    edges.push_back(fMin * std::pow(fMax / fMin, t));
  }
  return edges;
}

std::vector<int> logBandEdges(int sampleRateHz, int fftSize) {
  const std::vector<double> edgesHz = logBandEdgeHz(sampleRateHz);
  const double binHz = double(sampleRateHz) / double(fftSize);
  const int maxBin = (fftSize / 2) - 1;
  std::vector<int> bins;
  bins.reserve(edgesHz.size());
  for (double hz : edgesHz) {
    int bin = int(std::floor(hz / binHz));
    bins.push_back(std::clamp(bin, 1, maxBin));
  }
  // Independent clamping collapses the lowest log edges onto FFT bin 1 when
  // bin width (~43 Hz at 1024/44.1 kHz) is wider than the first bands, which
  // leaves bars 0–1 with lo==hi and dumps all bass into bar 2. Keep edges
  // strictly increasing so every bar owns at least one bin.
  for (size_t i = 1; i < bins.size(); ++i) {
    if (bins[i] <= bins[i - 1]) {
      bins[i] = bins[i - 1] + 1;
    }
  }
  if (bins.back() > maxBin) {
    bins.back() = maxBin;
    for (int i = int(bins.size()) - 2; i >= 0; --i) {
      if (bins[size_t(i)] >= bins[size_t(i + 1)]) {
        bins[size_t(i)] = bins[size_t(i + 1)] - 1;
      }
    }
  }
  return bins;
}

void fftInPlace(std::vector<double>& re, std::vector<double>& im) {
  const int n = int(re.size());
  if (n == 0 || (n & (n - 1)) != 0) {
    throw std::invalid_argument("fftSize must be power of two");
  }

  int j = 0;
  for (int i = 1; i < n; ++i) {
    int bit = n >> 1;
    for (; j & bit; bit >>= 1) {
      j ^= bit;
    }
    j ^= bit;
    if (i < j) {
      std::swap(re[size_t(i)], re[size_t(j)]);
      std::swap(im[size_t(i)], im[size_t(j)]);
    }
  }

  for (int len = 2; len <= n; len <<= 1) {
    const double ang = -2.0 * kPi / double(len);
    const double wlenRe = std::cos(ang);
    const double wlenIm = std::sin(ang);
    for (int i = 0; i < n; i += len) {
      double wRe = 1.0;
      double wIm = 0.0;
      for (int k = 0; k < len / 2; ++k) {
        const double uRe = re[size_t(i + k)];
        const double uIm = im[size_t(i + k)];
        const double vRe = re[size_t(i + k + len / 2)] * wRe - im[size_t(i + k + len / 2)] * wIm;
        const double vIm = re[size_t(i + k + len / 2)] * wIm + im[size_t(i + k + len / 2)] * wRe;
        re[size_t(i + k)] = uRe + vRe;
        im[size_t(i + k)] = uIm + vIm;
        re[size_t(i + k + len / 2)] = uRe - vRe;
        im[size_t(i + k + len / 2)] = uIm - vIm;
        const double nextWRe = wRe * wlenRe - wIm * wlenIm;
        wIm = wRe * wlenIm + wIm * wlenRe;
        wRe = nextWRe;
      }
    }
  }
}

}  // namespace

StftBandFolder::StftBandFolder(int fftSize, int framesPerSecond)
    : fftSize_(fftSize), framesPerSecond_(framesPerSecond) {}

QVector<std::array<double, AudioLevels::kBandCount>> StftBandFolder::analyze(
    const QVector<double>& samples, int sampleRateHz) const {
  QVector<std::array<double, AudioLevels::kBandCount>> frames;
  if (samples.isEmpty() || sampleRateHz <= 0) {
    frames.push_back({});
    return frames;
  }

  const int hop = std::max(1, sampleRateHz / framesPerSecond_);
  const std::vector<double> window = hann(fftSize_);
  const std::vector<int> bandEdges = logBandEdges(sampleRateHz, fftSize_);
  const int magCount = fftSize_ / 2;

  for (int start = 0; start + fftSize_ <= samples.size(); start += hop) {
    std::vector<double> re(size_t(fftSize_), 0.0);
    std::vector<double> im(size_t(fftSize_), 0.0);
    for (int i = 0; i < fftSize_; ++i) {
      re[size_t(i)] = samples[start + i] * window[size_t(i)];
    }
    fftInPlace(re, im);

    std::vector<double> mags(size_t(magCount), 0.0);
    for (int k = 0; k < magCount; ++k) {
      mags[size_t(k)] = std::sqrt(re[size_t(k)] * re[size_t(k)] + im[size_t(k)] * im[size_t(k)]);
    }

    std::array<double, AudioLevels::kBandCount> bands{};
    for (int b = 0; b < AudioLevels::kBandCount; ++b) {
      const int lo = bandEdges[size_t(b)];
      const int hi = bandEdges[size_t(b + 1)];
      double sum = 0;
      int count = 0;
      for (int k = lo; k < hi && k < magCount; ++k) {
        sum += mags[size_t(k)];
        ++count;
      }
      bands[size_t(b)] = count == 0 ? 0.0 : sum / double(count);
    }

    double peak = 0;
    for (double v : bands) {
      if (v > peak) peak = v;
    }
    if (peak > 1e-12) {
      for (double& v : bands) {
        v = std::clamp(v / peak, 0.0, 1.0);
      }
    }
    frames.push_back(bands);
  }

  if (frames.isEmpty()) {
    frames.push_back({});
  }
  return frames;
}

double StftBandFolder::bandCenterHz(int index, int sampleRateHz) {
  const std::vector<double> edgesHz = logBandEdgeHz(sampleRateHz);
  return std::sqrt(edgesHz[size_t(index)] * edgesHz[size_t(index + 1)]);
}

}  // namespace aoide
